README.md
Favicon Override plugin for Elgg 4.0 and newer Elgg 4.X
=======================================================

Latest Version: 4.0.0  
Released: 2022-05-22  
Contact: iionly@gmx.de  
License: GNU General Public License version 2  
Copyright: (C) iionly 2016


Description
-----------

This plugin allows you to override the default favicon of Elgg.


Installation
------------

1. If you have a previous version of the Favicon Override plugin installed, disable the plugin first and then remove the favicon_override plugin folder from the mod directory completely (don't forget to save your favicon images from the "mod/favicon_override/graphics" directory),
2. Copy the favicon_override plugin folder into the mod directory of your Elgg installation and place your favicon images (resolutions 16x16px, 32x32px, 64x64px and 128x128px in png, favicon.ico and favicon.svg) into the "mod/favicon_override/graphics" folder. Rename your images to the names of the existing placeholder images of the corresponding sizes and formats,
3. Enable the plugin in the admin section of your site. For your favicon image to show up immediately in your browser you might need to flush the site cache of your Elgg site and also the browser cache. Later on your favicon should show up automatically for your site users once the possibly cached files have expired (e.g. after 24h).


CHANGES.txt
Changelog
=========

4.0.0
-----

- Requires Elgg 4.0 at minimum,
- Updated for Elgg 4.

3.3.0
-----

- Requires Elgg 3.3 at minimum,
- Updated for Elgg 3.3.

3.0.0
-----

- Requires Elgg 3.0 at minimum,
- Updated for Elgg 3.

2.3.1
-----

- Requires Elgg 2.3 at minimum,
- Handling of favicon.svg added (you need to provide a suitable favicon image in svg format... there are online converters available all around the Internet...),
- Fixed serving of favicon.ico.

2.0.0
-----

- Initial release


lib/hooks.php
<?php

function favicon_override_head(\Elgg\Hook $hook) {
	$data = $hook->getValue();

	$data['links']['apple-touch-icon'] = [
		'rel' => 'apple-touch-icon',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-128.png'),
	];
	$data['links']['icon-ico'] = [
		'rel' => 'icon',
		'href' => elgg_get_simplecache_url('favicon_override/favicon.ico'),
	];
	$data['links']['icon-vector'] = [
		'rel' => 'icon',
		'sizes' => '16x16 32x32 48x48 64x64 128x128',
		'type' => 'image/svg+xml',
		'href' => elgg_get_simplecache_url('favicon_override/favicon.svg'),
	];
	$data['links']['icon-16'] = [
		'rel' => 'icon',
		'sizes' => '16x16',
		'type' => 'image/png',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-16.png'),
	];
	$data['links']['icon-32'] = [
		'rel' => 'icon',
		'sizes' => '32x32',
		'type' => 'image/png',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-32.png'),
	];
	$data['links']['icon-64'] = [
		'rel' => 'icon',
		'sizes' => '64x64',
		'type' => 'image/png',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-64.png'),
	];
	$data['links']['icon-128'] = [
		'rel' => 'icon',
		'sizes' => '128x128',
		'type' => 'image/png',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-128.png'),
	];

	return $data;
}


classes/FaviconOverrideHooks.php
<?php

class FaviconOverrideHooks {

	public static function favicon_override_head(\Elgg\Hook $hook) {
		$data = $hook->getValue();

		$data['links']['apple-touch-icon'] = [
			'rel' => 'apple-touch-icon',
			'href' => elgg_get_simplecache_url('favicon_override/favicon-128.png'),
		];
		$data['links']['icon-ico'] = [
			'rel' => 'icon',
			'href' => elgg_get_simplecache_url('favicon_override/favicon.ico'),
		];
		$data['links']['icon-vector'] = [
			'rel' => 'icon',
			'sizes' => '16x16 32x32 48x48 64x64 128x128',
			'type' => 'image/svg+xml',
			'href' => elgg_get_simplecache_url('favicon_override/favicon.svg'),
		];
		$data['links']['icon-16'] = [
			'rel' => 'icon',
			'sizes' => '16x16',
			'type' => 'image/png',
			'href' => elgg_get_simplecache_url('favicon_override/favicon-16.png'),
		];
		$data['links']['icon-32'] = [
			'rel' => 'icon',
			'sizes' => '32x32',
			'type' => 'image/png',
			'href' => elgg_get_simplecache_url('favicon_override/favicon-32.png'),
		];
		$data['links']['icon-64'] = [
			'rel' => 'icon',
			'sizes' => '64x64',
			'type' => 'image/png',
			'href' => elgg_get_simplecache_url('favicon_override/favicon-64.png'),
		];
		$data['links']['icon-128'] = [
			'rel' => 'icon',
			'sizes' => '128x128',
			'type' => 'image/png',
			'href' => elgg_get_simplecache_url('favicon_override/favicon-128.png'),
		];

		return $data;
	}
}

start.php
<?php

elgg_register_event_handler('init', 'system', 'favicon_override_init');

function favicon_override_init() {
	elgg_register_plugin_hook_handler('head', 'page', 'favicon_override_head');
}

function favicon_override_head(\Elgg\Hook $hook) {
	$data = $hook->getValue();

	$data['links']['apple-touch-icon'] = [
		'rel' => 'apple-touch-icon',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-128.png'),
	];
	$data['links']['icon-ico'] = [
		'rel' => 'icon',
		'href' => elgg_get_simplecache_url('favicon_override/favicon.ico'),
	];
	$data['links']['icon-vector'] = [
		'rel' => 'icon',
		'sizes' => '16x16 32x32 48x48 64x64 128x128',
		'type' => 'image/svg+xml',
		'href' => elgg_get_simplecache_url('favicon_override/favicon.svg'),
	];
	$data['links']['icon-16'] = [
		'rel' => 'icon',
		'sizes' => '16x16',
		'type' => 'image/png',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-16.png'),
	];
	$data['links']['icon-32'] = [
		'rel' => 'icon',
		'sizes' => '32x32',
		'type' => 'image/png',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-32.png'),
	];
	$data['links']['icon-64'] = [
		'rel' => 'icon',
		'sizes' => '64x64',
		'type' => 'image/png',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-64.png'),
	];
	$data['links']['icon-128'] = [
		'rel' => 'icon',
		'sizes' => '128x128',
		'type' => 'image/png',
		'href' => elgg_get_simplecache_url('favicon_override/favicon-128.png'),
	];

	return $data;
}
